File: Algorithm/may_2024/may_16_2024/may_16_2024.js
/* 
Let's define a "sevenish" number to be one which is either a power of 7, or the sum of unique powers of 7. The first few sevenish numbers are 1, 7, 8, 49, and so on. Create an algorithm to find the nth sevenish number.
*/

let currentPotency = 1;
const sevenishList = [1];

function getSevenishNumber(num){
    if(num <= 0){
        console.error("Number needs to be higher than 0");
        return undefined;
    }
    return populateSevenishList(num)
}

function populateSevenishList(num){
	if(sevenishList.length === num){
  	return sevenishList[num-1];
  	}
  	sevenishList.push(7 ** currentPotency);
  	addExtras(0, num);
    currentPotency++;
    return populateSevenishList(num);
}

function addExtras(potencyToAdd, num){
    if(potencyToAdd === currentPotency || sevenishList.length === num){
    	 return;
    }
    sevenishList.push((7 ** currentPotency) + (7 ** potencyToAdd));
    potencyToAdd++;
    return addExtras(potencyToAdd, num); 
}

console.log(getSevenishNumber(7));
console.log(sevenishList);

File: Algorithm/may_2024/may_16_2024/may_16_2024.java
import java.util.ArrayList;
import java.util.List;

public class SevenishNumbers {

    private int currentPotency = 0;
    private List<Integer> sevenishList = new ArrayList<>();

    public int getSevenishNumber(int num) {
        if (num <= 0) {
            System.err.println("Number needs to be higher than 0");
            return -1; 
        }
        return populateSevenishList(num);
    }

    private int populateSevenishList(int num) {
        if (sevenishList.size() >= num) {
            return sevenishList.get(num - 1);
        }
        int nextSevenish = (int) Math.pow(7, (currentPotency));
        sevenishList.add(nextSevenish);
        if(currentPotency > 0){
            addExtras(0, num);
        }
        currentPotency++;
        return populateSevenishList(num);
    }

    private void addExtras(int potencyToAdd, int num) {
        if (potencyToAdd == currentPotency || sevenishList.size() >= num) {
            return;
        }
        int nextExtra = (int) (Math.pow(7, currentPotency) + Math.pow(7, potencyToAdd));
        sevenishList.add(nextExtra);
        potencyToAdd++;
        addExtras(potencyToAdd, num);
    }

    public static void main(String[] args) {
        SevenishNumbers sevenishNumbers = new SevenishNumbers();
        System.out.println(sevenishNumbers.getSevenishNumber(7));
        System.out.println(sevenishNumbers.sevenishList);
    }

}